hw/mcu/nordic/nrf5340/src/hal_timer.c

```diff
@@ -0,0 +1,925 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include <stdint.h>
+#include <assert.h>
+#include <errno.h>
+#include "os/mynewt.h"
+#include "mcu/cmsis_nvic.h"
+#include "hal/hal_timer.h"
+#include "nrf.h"
+#include "mcu/nrf5340_hal.h"
+
+/* IRQ prototype */
+typedef void (*hal_timer_irq_handler_t)(void);
+
+/* User CC 2 for reading counter, CC 3 for timer isr */
+#define NRF_TIMER_CC_READ       (2)
+#define NRF_TIMER_CC_INT        (3)
+
+/* Output compare 2 used for RTC timers */
+#define NRF_RTC_TIMER_CC_INT    (2)
+
+/* Maximum number of hal timers used */
+#define nrf5340_hal_timer_MAX     (6)
+
+/* Maximum timer frequency */
+#define NRF5340_MAX_TIMER_FREQ    (16000000)
+
+struct nrf5340_hal_timer {
+    uint8_t tmr_enabled;
+    uint8_t tmr_irq_num;
+    uint8_t tmr_rtc;
+    uint8_t tmr_pad;
+    uint32_t tmr_cntr;
+    uint32_t timer_isrs;
+    uint32_t tmr_freq;
+    void *tmr_reg;
+    TAILQ_HEAD(hal_timer_qhead, hal_timer) hal_timer_q;
+};
+
+#if MYNEWT_VAL(TIMER_0)
+struct nrf5340_hal_timer nrf5340_hal_timer0;
+#endif
+#if MYNEWT_VAL(TIMER_1)
+struct nrf5340_hal_timer nrf5340_hal_timer1;
+#endif
+#if MYNEWT_VAL(TIMER_2)
+struct nrf5340_hal_timer nrf5340_hal_timer2;
+#endif
+#if MYNEWT_VAL(TIMER_3)
+struct nrf5340_hal_timer nrf5340_hal_timer3;
+#endif
+#if MYNEWT_VAL(TIMER_4)
+struct nrf5340_hal_timer nrf5340_hal_timer4;
+#endif
+
+static const struct nrf5340_hal_timer *nrf5340_hal_timers[nrf5340_hal_timer_MAX] = {
+#if MYNEWT_VAL(TIMER_0)
+    &nrf5340_hal_timer0,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_1)
+    &nrf5340_hal_timer1,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_2)
+    &nrf5340_hal_timer2,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_3)
+    &nrf5340_hal_timer3,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(TIMER_4)
+    &nrf5340_hal_timer4,
+#else
+    NULL,
+#endif
+};
+
+/* Resolve timer number into timer structure */
+#define nrf5340_hal_timer_RESOLVE(__n, __v)       \
+    if ((__n) >= nrf5340_hal_timer_MAX) {         \
+        rc = EINVAL;                            \
+        goto err;                               \
+    }                                           \
+    (__v) = (struct nrf5340_hal_timer *) nrf5340_hal_timers[(__n)];            \
+    if ((__v) == NULL) {                        \
+        rc = EINVAL;                            \
+        goto err;                               \
+    }
+
+/* Interrupt mask for interrupt enable/clear */
+#define NRF_TIMER_INT_MASK(x)    ((1 << (uint32_t)(x)) << 16)
+
+static uint32_t
+nrf_read_timer_cntr(NRF_TIMER_Type *hwtimer)
+{
+    uint32_t tcntr;
+
+    /* Force a capture of the timer into 'cntr' capture channel; read it */
+    hwtimer->TASKS_CAPTURE[NRF_TIMER_CC_READ] = 1;
+    tcntr = hwtimer->CC[NRF_TIMER_CC_READ];
+
+    return tcntr;
+}
+
+/**
+ * nrf timer set ocmp
```

Comment:
In all cases starting from here first line of comment repeats function name with underscores changed to spaces.
It does not provide any useful information and could be safely removed.